junto-rust/zomes/core/code/src/expressions/group.rs

```diff
@@ -0,0 +1,169 @@
+    let current_user_username = user::get_user_username_address_by_agent_address()?;
+    let group = user::get_user_pack(current_user_username)?.pack;   
+    match group {
+        Some(group_obj) => Ok(add_member_to_group(username_address.clone(), group_obj.address.clone())?),
+        None => Err(ZomeApiError::from("No pack entries on current user".to_string()))
+    }
+}
+
+pub fn add_member_to_group(username_address: Address, group: Address) -> ZomeApiResult<JsonString>{
+    let current_user_username = user::get_user_username_address_by_agent_address()?;
+    let group_owner = is_group_owner(group.clone(), current_user_username.clone())?;
+    if group_owner == true{
+        let group_member = is_group_member(group.clone(), username_address.clone())?;
+        if group_member == false{
+            hdk::api::link_entries(&group, &username_address, "member".to_string())?;
+            Ok(json!({ "message": "User added to group" }).into())
+        } else {
+            Err(ZomeApiError::from("User submitted is already a member of given group".to_string()))
+        }
+    } else {
+        Err(ZomeApiError::from("You are not the owner of given group and thus cannot ad/remove members".to_string()))
+    }
+}
+
+pub fn remove_group_member(username_address: Address, group: Address) -> ZomeApiResult<JsonString>{
+    let current_user_username = user::get_user_username_address_by_agent_address()?;
+    let group_owner = is_group_owner(group.clone(), current_user_username.clone())?;
+    if group_owner == true{
+        let group_member = is_group_member(group.clone(), username_address.clone())?;
+        if group_member == true{
+            hdk::api::remove_link(&group, &username_address, "member".to_string())?;
+            Ok(json!({ "message": "User removed from group" }).into())
+        } else {
+            Err(ZomeApiError::from("User submitted is not a group member of given group".to_string()))
+        }
+    } else {
+        Err(ZomeApiError::from("You are not the owner of given group and thus cannot ad/remove members".to_string()))
+    }
+}
+
+pub fn is_group_member(group: Address, user: Address) -> ZomeApiResult<bool>{
+    let group_entry = hdk::api::get_entry(&group)?;
+    match group_entry {
+        Some(Entry::App(_, entry_value)) => {
+            match app_definitions::Group::try_from(&entry_value){
+                Ok(_entry) => {
+                    match utils::get_links_and_load_type::<String, app_definitions::UserName>(&group, "member".to_string()){
+                        Ok(member_vec) => {
+                            for member in member_vec {
+                                if member.address == user{
+                                    return Ok(true)
+                                };
+                            };
+                            return Ok(false)
+                        },
+                        Err(err) => return Err(err)
```

Comment:
This one is not :P. Early returns seem to be a bit of an anti-pattern in Rust. I can see how this could be written with a lot less rightward drift using `?` instead